xmtp_mls/README.md

```diff
@@ -0,0 +1,136 @@
+}
+
+-- Used to keep track of the last seen message timestamp in a topic
+CREATE TABLE topic_refresh_state {
+    topic TEXT PRIMARY KEY NOT NULL,
+    last_message_timestamp BIGINT NOT NULL
+}
+
+-- This table is required to retry messages that do not send successfully due to epoch conflicts
+CREATE TABLE group_intents {
+    id SERIAL PRIMARY KEY NOT NULL,
+    kind INT NOT NULL, -- INTENT_KIND
+    group_id TEXT NOT NULL,
+    "data" BLOB NOT NULL, -- Some sort of serializable blob that can be used to re-try the message if the first attempt failed due to conflict
+    "state" INT NOT NULL, -- INTENT_STATE,
+    message_hash TEXT -- The hash of the encrypted, concrete, form of the message if it was published.
+}
+
+CREATE TABLE pending_outbound_welcome_messages {
+    id TEXT PRIMARY KEY NOT NULL, -- Derived via SHA256(CONCAT(group_id, welcome_message, ratchet_tree, installation_id))
+    installation_id TEXT PRIMARY KEY NOT NULL,
+    commit_hash TEXT NOT NULL, -- The hash of the commit which created this
+    group_id TEXT NOT NULL,
+    welcome_message BLOB NOT NULL,
+    ratchet_tree BLOB NOT NULL,
+}
+
+-- OpenMLS Keystore backing database to lookup objects. Interface and matching types defined here https://github.com/openmls/openmls/blob/main/traits/src/key_store.rs
+CREATE TABLE openmls_keystore {
+    "key" BLOB PRIMARY KEY NOT NULL,
+    "value" BLOB NOT NULL,
+}
+```
+
+## Enums
+
+### GROUP_MEMBERSHIP_STATE
+
+- ALLOWED // User has agreed to be a member of the group
+- REJECTED // User has rejected an invite to the group or left
+- PENDING // User has neither accepted or rejected whether they should join the group
+
+### INTENT_STATE
+
+- TO_SEND // Either has never been sent to the network or needs to be re-sent
+- PUBLISHED // Sent to the network but has not been read back or committed
+- COMMITTED // Committed messages could be deleted
+
+### INTENT_KIND
+
+- APPLICATION // An intent to send a message to the group
+- ADD_MEMBERS // An intent to add members to the group
+- REMOVE_MEMBERS // An intent to remove members from the group
+- KEY_UPDATE // An intent to update your own group key
+
+## State Machine
+
+The [following diagram](https://app.excalidraw.com/s/4nwb0c8ork7/6pPH1kQDoj3) illustrates some common flows in the state machine
+
+![MLS State Machione](../img/mls-state-machine.png "MLS State Machine")
+
+### Known missing items from the state machine
+
+- Key updates
+- Processing incoming welcome messages
+
+### Add members to a group
+
+Simplified high level flow for adding members to a group:
+
+1. Consume Key Packages for all new members
+1. Create a `group_intent` for adding the members
+1. Sync the state of the group with the network
+1. Convert the intent into concrete commit and welcome messages for the current epoch
+1. Publish commit message
+1. Sync the state of the group with the network
+1. If no conflicts: Publish welcome messages to new members.
+   If conflicts: Go back to step 4 and try again
```

Comment:
So at a high level all of these actions involve publishing a payload, and then only applying the change once you read the payload back from the network and confirm it has been placed in-order without any conflicts with other changes. We're essentially placing actions onto a queue.

Once you read the payload back from the queue, you need some kind of mechanism to correlate the payload back to your intent, and apply any remaining actions associated with the intent (e.g. send out welcome messages). 

Ideally:
1. Any data needed to correlate the payload back to the intent is in the database, rather than held in-memory (in the event that the app is killed before the payload is read).
2. The whole process should not block the user from taking further actions, as it involves multiple network round-trips. For example, you could add a member and then send a message (before receiving a response from the network), and then apply both changes when you read the corresponding payloads.

I'm wondering if we can support this kind of asynchronicity and robustness in the face of unreliable network conditions and randomly killed app lifetimes by splitting these into smaller atomic steps roughly mapping to the `intent_state` you enumerated above. Each of these steps can be picked up at any time (e.g. after an app crash, after coming back online) and don't block on any other step:

**1. User action loop**
User actions can be thought of as an ordered stream of events - for each user action we create a corresponding intent with `TO_SEND` state and store it in the database. The intents MUST be ordered in the database.

**2. Intent publishing loop**
Read through each `TO_SEND` intent in order and map it into the corresponding MLS messages (including downloading key packages if necessary). Publish them onto the network, and then once published update the state to `PUBLISHED` and move onto the next intent.

You may also need to persist some state, such as welcome messages, onto the intents table so that it can be retrieved later - perhaps there are separate `publish_data` and `commit_data` columns.

**3. Payload receiving loop**
Read payloads from the network in-order. Most of them will be messages or payloads from other installations. For payloads from your own installation, map them back to your intent, apply any remaining actions, and then update the intent state to `COMMITTED`. This loop can be a pull-based loop or a push-based loop, it doesn't matter.

These three loops can be running in three separate threads, or they can be methods that are manually invoked at regular intervals or on-demand from the same thread, it doesn't matter.

Reply:
I like where you're going with this. Making things as parallelizable and re-entrant as possible is a worthwhile goal.

You want to try and fork the Excalidraw and see what these changes would look like in the state machine?